ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Keystore.hs

```diff
@@ -153,47 +106,27 @@ bracketTestKeystore withKeystore =
 -- races due to the fact its underlying file is stored in the OS' temporary
 -- directory.
 newTestKeystore :: IO Keystore
-newTestKeystore = liftIO $ runIdentityT $ fromKeystore $ do
-    fp <- liftIO $ getKeystorePath
-    us <- peekUserSecret fp
-    Keystore <$> newMVar (InternalStorage us)
-  where
-    -- | Generates a new path for the temporary 'Keystore'.
-    -- Unfortunately, System.IO does not expose an API for path
-    -- generation (it is implemented in `openTempFile'`), so
-    -- we create the file, then close the handle and remove it.
-    -- It cannot be left empty since the `instance Bi` for
-    -- 'UserSecret' cannot handle empty files.
-    getKeystorePath :: IO FilePath
-    getKeystorePath = do
-        tempDir <- getTemporaryDirectory
-        let dir = tempDir </> "ariadne-test"
-        createDirectoryIfMissing False dir
-        (fp, hdl) <- openTempFile dir "keystore.key"
-        hClose hdl
-        removeFile fp
-        pure fp
+newTestKeystore = do
+    fp <- liftIO getTempKeystorePath
+    istorage <- peekInternalStorage fp
+    Keystore <$> newMVar istorage
 
 -- | Release the resources associated with this 'Keystore'.
 releaseKeystore :: DeletePolicy -> Keystore -> IO ()
 releaseKeystore dp (Keystore ks) =
     -- We are not modifying the 'MVar' content, because this function is
     -- not exported and called exactly once from the bracket de-allocation.
-    withMVar ks $ \internalStorage@(InternalStorage us) -> do
-        fp <- release internalStorage
+    withMVar ks $ \istorage -> do
+        fp <- release istorage
         case dp of
              KeepKeystoreIfEmpty   -> pass
              RemoveKeystoreIfEmpty ->
-                 when (isEmptyUserSecret us) $ removeFile fp
+                 when (isEmptyInternalStorage istorage) $ removeFile fp
 
 -- | Releases the underlying 'InternalStorage' and returns the updated
 -- 'InternalStorage' and the file on disk this storage lives in.
--- 'FilePath'.
 release :: InternalStorage -> IO FilePath
-release (InternalStorage us) = do
-    let fp = getUSPath us
-    writeUserSecret us
-    return fp
+release istorage = writeInternalStorage istorage $> istorage  ^. isFilePath
```

Comment:
It seems to be the only place where `isFilePath` is actually used. And that can be easily avoided. `release` is only called from `releaseKeystore`. It's called from `keystoreComponent`, `bracketKeystore` and `bracketTestKeystore`. The first two can pass `FilePath` to `releaseKeystore` already. `bracketTestKeystore` needs to be changed a little bit so that `newTestKeystore` returns a pair.

I don't understand this design when `FilePath` is part of `Keystore`. I know it was like that before, but I'd like to get rid of that unless someone provides a good reason why it's good.

One reason why it's bad is that when we parse `JSON` we return `Keystore` with some static (empty) file path. That's already enough for me to dislike this design.

Reply:
@gromakovsky The reason why we have filePath as part of `keystore` is that it's passed as a configuration value from ` WalletConfig` in `wcKeyfilePath`

if we derive the filePath from within `Ariadne.Wallet.Cardano.Kernel.Keystore.Persistence` then we can get rid of filePath in `keystore`.

Reply:
@gromakovsky The reason is that it allows us to use keystore operations such as `insert` and `delete` without use of a `filePath` value in function arguments.

If we derived filePath from with in `Ariadne.Wallet.Cardano.Kernel.Keystore.Persistence` then it would be easy to get rid of filePath from with in `keystore`.

Reply:
Ok, I agree, I didn't notice that. However, instead of storing `FilePath` in `InternalStorage`, why not store it in `Keystore`? This way we won't have to set default value in JSON deserialization.

Reply:
That sounds good.